Src/WindowsServer/WindowsServer.Shared/Implementation/AzureMetadataRequestor.cs

```diff
@@ -0,0 +1,109 @@
+        /// </summary>
+        /// parameter sent to the func is a string representing the Uri to request Azure IMS data from.
+        /// <returns>An instance of AzureInstanceComputeMetadata or null.</returns>
+        private Func<string, Task<AzureInstanceComputeMetadata>> azureIMSRequestor = null;
+
+        internal AzureMetadataRequestor(Func<string, Task<AzureInstanceComputeMetadata>> makeAzureIMSRequestor = null)
+        {
+            this.azureIMSRequestor = makeAzureIMSRequestor;
+        }
+
+        /// <summary>
+        /// Gets or sets the base URI for the Azure Instance Metadata service. Internal to allow overriding in test.
+        /// </summary>
+        internal string BaseAimsUri { get; set; } = "http://169.254.169.254/metadata/instance/compute";
+
+        public async Task<AzureInstanceComputeMetadata> GetAzureComputeMetadataAsync()
+        {
+            string metadataRequestUrl = $"{this.BaseAimsUri}?{AzureMetadataRequestor.AzureImsJsonFormat}&{AzureMetadataRequestor.AzureImsApiVersion}";
+
+            AzureInstanceComputeMetadata jsonResponse = await this.MakeAzureMetadataRequestAsync(metadataRequestUrl);
+
+            return jsonResponse;
+        }
+
+        private async Task<AzureInstanceComputeMetadata> MakeAzureMetadataRequestAsync(string metadataRequestUrl)
+        {
+            AzureInstanceComputeMetadata requestResult = null;
+
+            SdkInternalOperationsMonitor.Enter();
+            try
+            {
+                if (this.azureIMSRequestor != null)
+                {
+                    requestResult = await this.azureIMSRequestor(metadataRequestUrl);
+                }
+                else
+                {
+                    requestResult = await this.MakeWebRequestAsync(metadataRequestUrl);
+                }
+            }
+            catch (Exception ex)
+            {
+                WindowsServerEventSource.Log.AzureInstanceMetadataRequestFailure(
+                    metadataRequestUrl, ex.Message, ex.InnerException != null ? ex.InnerException.Message : string.Empty);
+            }
+            finally
+            {
+                SdkInternalOperationsMonitor.Exit();
+            }
+
+            return requestResult;
+        }
+        
+        private async Task<AzureInstanceComputeMetadata> MakeWebRequestAsync(string requestUrl)
+        {
+            AzureInstanceComputeMetadata azureIms = null;
+            DataContractJsonSerializer deserializer = new DataContractJsonSerializer(typeof(AzureInstanceComputeMetadata));
+
+            using (var azureImsClient = new HttpClient())
+            {
+                azureImsClient.MaxResponseContentBufferSize = AzureMetadataRequestor.AzureImsMaxResponseBufferSize;
+                azureImsClient.DefaultRequestHeaders.Add("Metadata", "True");
+                azureImsClient.Timeout = this.AzureImsRequestTimeout;
+
+                Stream content = await azureImsClient.GetStreamAsync(requestUrl);
```

Comment:
Two things:
1. `.ConfigureAwait(false);`
2. Dispose of `content` (assuming the deserializer doesn't already close the stream)

Reply:
Again, thanks. I am not clear on the necessity of calling `Dispose` explicitly though - would have thought that the Stream going out of scope would take care of it? Perhaps there is a benefit to performance/memory usage here?